keps/sig-apps/20190318-Graduate-CronJob-to-Stable.md

```diff
@@ -0,0 +1,552 @@
+---
+title: graduate-cronjob-to-stable  
+authors:  
+  - "@barney-s"  
+
+owning-sig: sig-apps  
+participating-sigs:
+  - sig-scaling
+
+reviewers:
+  - "@liggitt"
+  - "@kow3ns"
+  - "@janetkuo"
+  - “@mortent”
+
+approvers:
+  - "@kow3ns"
+  - "@liggitt"
+  - "@soltysh"
+
+editor: TBD  
+creation-date: 2019-04-18  
+last-updated: 2019-04-18  
+status: provisional  
+see-also:  
+replaces:  
+superseded-by:  
+
+---
+
+# Graduate CronJob to stable
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+- [Goals](#goals)
+- [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [Promote CronJob API to v1](#promote-cronjob-api-to-v1)
+  - [Rearchitect CronJob controller](#rearchitect-cronjob-controller)
+    - [Existing controller](#existing-controller)
+    - [Proposed rearchitect](#proposed-rearchitect)
+      - [Informers and Caches](#informers-and-caches)
+      - [Multiple workers](#multiple-workers)
+      - [Handling Cron aspect](#handling-cron-aspect)
+      - [Metrics](#metrics)
+  - [Add .status.lastSuccessfulTime](#add-statuslastsuccessfultime)
+  - [Add .status.nextScheduleTime](#add-statusnextscheduletime)
+  - [Add Counters](#add-counters)
+  - [Add .status.conditions](#add-statusconditions)
+  - [CatchUp concurrency support](#catchup-concurrency-support)
+  - [Use master timezone instead of UTC](#use-master-timezone-instead-of-utc)
+  - [Fix applicable open issues](#fix-applicable-open-issues)
+  - [Scale Targets for GA](#scale-targets-for-ga)
+    - [CronJob Limits](#cronjob-limits)
+    - [Frequency of launched jobs](#frequency-of-launched-jobs)
+- [V1 API](#v1-api)
+- [v1beta1 changes](#v1beta1-changes)
+- [Validations](#validations)
+- [Tests](#tests)
+  - [E2E test](#e2e-test)
+    - [Existing test cases](#existing-test-cases)
+    - [New test cases](#new-test-cases)
+  - [Conformance Tests](#conformance-tests)
+  - [Unit Tests](#unit-tests)
+- [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+<!-- /toc -->
+
+## Summary
+
+[CronJob](https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/) is a Kubernetes API that creates Job object on a schedule specified by a cron spec. It is in beta status since v1.8. This document lays out the plan to promote it to stable.
+
+## Motivation
+
+CronJob definition has been stable for the last few releases and is useful to run periodic tasks in kubernetes cluster. This API adds the ability to add cron facility to a cluster. We feel the API is ready to be promoted to Stable and be supported longterm by the community.
+
+## Goals
+
+- Rearchitect CronJob controller
+- .status.lastSuccessfulTime
+- .status.nextRunTime
+- Use master timezone instead of UTC
+- Fix open issues and add tests
```

Comment:
The goals as I see it are:
- write a new controller which will:
  - use informers rather than polling
  - address open issues and problems with the current controller
  - add metrics exposing controller throughput, state (others?)
- extend CronJob status field with:
  - lastSuccesfulTime - (explain here what it's for)
  - nextRunTime - (explain here what it's for)

Reply:
thanks for rewording